### PR TITLE
Simplify shortvec optimization

### PR DESCRIPTION
**Pull request by garious**
#### Problem

The serialization of Transaction requires a bunch of custom code to be added when new fields are added to it. But without the shortvec optimization, the Transaction and Instruction modules are using 8 bytes to encode the length of each vector when typically only 1 is needed (`Instruction::data` may require 2).

#### Summary of Changes

This PR builds on https://github.com/solana-labs/solana/pull/3471. Rather than merge that PR, we're going to stay off on a branch until we're sure this change won't degrade singlenode performance.

* Remove the original shortvec optimization
* Demonstrate `#[serde(with = "short_vec")]` on `Vec<T>` to do a multi-byte encoding of each vector's length.

https://serde.rs/variant-attrs.html

TODO:

- [x] Implement deserialize()
- [x] Un-ignore tests that require deserialize()
- [x] Add tests
- [x] Benchmark before and after

### CODE REVIEW

File: sdk/src/short_vec.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
+use serde::de::{self, Deserializer, SeqAccess, Visitor};
+use serde::ser::{SerializeTuple, Serializer};
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use std::marker::PhantomData;
+use std::mem::size_of;
+
+/// Same as usize, but serialized with 1 to 9 bytes. If the value
+/// is above 0x7f, the top bit is set and the remaining value is
+/// stored in the follow byte. The following byte follows the same
```
Comment (CriesofCarrots) [inline sdk/src/short_vec.rs]: follow -> following

File: sdk/src/short_vec.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
+use serde::de::{self, Deserializer, SeqAccess, Visitor};
+use serde::ser::{SerializeTuple, Serializer};
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use std::marker::PhantomData;
+use std::mem::size_of;
+
+/// Same as usize, but serialized with 1 to 9 bytes. If the value is above
+/// 0x7f, the top bit is set and the remaining value is stored in the next
+/// byte. The next byte follows the same pattern until the 9th byte, which
```
Comment (CriesofCarrots) [inline sdk/src/short_vec.rs]: On second read: "Each byte follows the same pattern..." ?

File: core/src/sigverify.rs [hunk 3]
```diff
@@ -124,23 +121,18 @@ pub fn ed25519_verify(batches: &[SharedPackets]) -> Vec<Vec<u8>> {
...
+    let (_pubkey_len, pubkey_size) = decode_len(&packet.data[msg_start_offset..]);
+    let pubkey_offset = msg_start_offset + pubkey_size;
+
+    let sig_start = current_offset as usize + sig_size;
 
     (
         sig_len as u32,
         sig_start as u32,
-        current_offset + msg_start_offset as u32,
+        msg_start_offset as u32,
```
Comment (garious) [inline core/src/sigverify.rs]: @sakridge, there doesn't seem to be any test that checks this value. I can add `current_offset` or not right here, and all tests pass.  But upon closer inspection, I don't think there was a bug in the original code. And I don't *think* I added one.

File: core/src/sigverify.rs [hunk 4]
```diff
@@ -422,9 +414,9 @@ mod tests {
         let (sig_len, sig_start, msg_start_offset, pubkey_offset) =
             sigverify::get_packet_offsets(&packet, 0);
```
Comment (garious) [inline core/src/sigverify.rs]: The only test passes in zero.  My assumption is that if I passed in 10, we should add 10 to all outputs except `sig_len`?

File: core/src/sigverify.rs (hunk 3 above)
Comment (sakridge) [reply core/src/sigverify.rs]: We should add a test that a vector with multiple different transactions works.

Comment (garious) [reply core/src/sigverify.rs]: Yeah, one with multiple signatures would make it less fragile too.

File: core/src/sigverify.rs (hunk 4 above)
Comment (sakridge) [reply core/src/sigverify.rs]: believe so

File: core/src/sigverify.rs (hunk 3 above)
Comment (sakridge) [reply core/src/sigverify.rs]: ok, are you going to add back the add?

Comment (garious) [reply core/src/sigverify.rs]: I don't think I need to. I'm expanding the original test suite in an upcoming PR to be sure.

**Comment by garious** [comment]
doh, broke something that isn't evident until `test_blockstream_service_process_entries`. @CriesofCarrots, any thoughts offhand of what might be sensitive to the transaction's binary format?

**Comment by CriesofCarrots** [comment]
@garious , based on the line number of the panic, it looks like `serde_json::from_str` is failing. It might be related to this comment: https://github.com/solana-labs/solana/blob/845ddc349695427359efca546567086dddd113e4/core/src/blockstream_service.rs#L186
but I'm not sure what about your changes is causing the issue to show up earlier.
I'll pull your branch and look more closely

File: core/src/sigverify.rs (hunk 3 above)
Comment (garious) [reply core/src/sigverify.rs]: yep, all good

**Comment by garious** [comment]
@CriesofCarrots, it's my abuse of `serialize_tuple`. I need to rework it to get serde_json to generate valid json:

```
#[test]
    fn test_short_vec_json() {
        let vec = ShortVec(vec![0u8]);
        let s = serde_json::to_string(&vec).unwrap();
        assert_eq!(s, "[],[],1],0]");
    }
```

**Comment by CriesofCarrots** [comment]
@garious , excellent, looks like valid json now. However, `serde_json::to_string(&Entry)` used to serialize the transactions field into our binary format (`Array(Array(u8))`). Now it preserves the tx structure (`Array(Object(<tx fields>))`), eg:
```
"transactions": [{"signatures":[[1],[144,159,229,233,42,92,154,9,242,87,250,148,178,60,92,183,145,9,57,200,244,2,97,102,203,104,120,79,191,162,99,229,86,255,67,185,89,229,71,98,90,241,85,239,84,105,32,145,184,162,13,13,189,37,152,254,130,152,151,153,143,59,162,9]],"account_keys":[[1],[116,155,255,12,129,171,71,126,178,64,135,249,47,234,217,216,27,73,54,217,224,199,98,62,88,21,151,34,186,130,85,227]],"recent_blockhash":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"fee":0,"program_ids":[[1],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]],"instructions":[[1],{"program_ids_index":0,"accounts":[[2],0,0],"data":[[52],0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}]}]
```

Not sure how hard it will be to recover this in your serde work. One option is to patch blockstream to manually serialize the txs before streaming...

File: core/src/sigverify.rs (hunk 3 above)
Comment (sakridge) [reply core/src/sigverify.rs]: test_verify_seventy_one with --features=cuda enabled catches this. The CPU path doesn't use the current_offset (always passes 0), but the GPU path does.

**Comment by garious** [comment]
> I think this should cause test_verify_seventy_one to fail with the cuda build enabled.

@sakridge, all CI builds pass. Is that test enabled?

**Comment by sakridge** [comment]
@garious it was disabled by the move to virtual manifest that also broke the other stuff. so the perf build is not actually running with the flags enabled.

**Comment by garious** [comment]
> @garious , excellent, looks like valid json now. However, `serde_json::to_string(&Entry)` used to serialize the Entry transactions field into our binary format (`Array(Array(u8))`). Now it preserves the tx structure (`Array(Object(<tx fields>))`), eg:
> 
> ```
> "transactions": [{"signatures":[[1],[144,159,229,233,42,92,154,9,242,87,250,148,178,60,92,183,145,9,57,200,244,2,97,102,203,104,120,79,191,162,99,229,86,255,67,185,89,229,71,98,90,241,85,239,84,105,32,145,184,162,13,13,189,37,152,254,130,152,151,153,143,59,162,9]],"account_keys":[[1],[116,155,255,12,129,171,71,126,178,64,135,249,47,234,217,216,27,73,54,217,224,199,98,62,88,21,151,34,186,130,85,227]],"recent_blockhash":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"fee":0,"program_ids":[[1],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]],"instructions":[[1],{"program_ids_index":0,"accounts":[[2],0,0],"data":[[52],0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}]}]
> ```
> Not sure how hard it would be to recover this in your serde work. One option is to patch blockstream to manually serialize the txs before streaming...

@CriesofCarrots, the way transaction was double-serialized was not the intention of the shortvec optimization. The only thing weird about this new version is the multi-byte encoding of the array length in the first item of every JSON-encoded Vec.  To get rid of that, we have a few options:

1. We can fork bincode and change `serialize_seq()` to not `serialize_u64(len)`
2. We can change ShortVec to serialize a struct instead of a tuple, so you'll see `{"len": [1], "data": [...]}`in JSON instead of `[[1], ...]`.
3. Blockstream can map `Transaction` to `bincode::serialize(&transaction)`
4. Blockstream can use abandon `serde_json` in favor of `bincode` and maybe some client-side WebAssembly, which might turn out to be the only way Blockexplorer will be able to handle 100k+ TPS anyway.

**Comment by garious** [comment]
> I think this should cause test_verify_seventy_one to fail with the cuda build enabled.

Are you able to update sigverify.rs with whatever constraint that test depends on? Does that test actually fail or do you just think it might? Can you offer any more information that might help me understand why you think it would fail?

File: core/src/sigverify.rs (hunk 3 above)
Comment (garious) [reply core/src/sigverify.rs]: Did you see the test I added and rebased this PR on? Also, did you see how `current_offset` is now used a few lines prior? https://github.com/solana-labs/solana/pull/3545/files#diff-ad695894f5a727f4ce60e68987c4e35aR125

**Comment by sakridge** [comment]
@garious it fails in my local machine with that change.

**Comment by CriesofCarrots** [comment]
> @CriesofCarrots, the way transaction was double-serialized was not the intention of the shortvec optimization. The only thing weird about this new version is the multi-byte encoding of the array length in the first item of every JSON-encoded Vec. To get rid of that, we have a few options:
> 
>     1. We can fork bincode and change `serialize_seq()` to not `serialize_u64(len)`
> 
>     2. We can change ShortVec to serialize a struct instead of a tuple, so you'll see `{"len": [1], "data": [...]}`in JSON instead of `[[1], ...]`.
> 
>     3. Blockstream can map `Transaction` to `bincode::serialize(&transaction)`
> 
>     4. Blockstream can use abandon `serde_json` in favor of `bincode` and maybe some client-side WebAssembly, which might turn out to be the only way Blockexplorer will be able to handle 100k+ TPS anyway.

@garious , I think (3) is easy enough to implement right now, and prevents a breaking change in blockexplorer. But fair point on (4), I'll start thinking about that as I write something toward (3).

**Comment by garious** [comment]
> @garious it fails in my local machine with that change.

What is the intent of that test? What is the significance of 71?

**Comment by sakridge** [comment]
@garious Intent is just to verify the signature verify path with a case that hits the GPU. 71 is just the number of the transactions in the vector it checks.

**Comment by sakridge** [comment]
#3556 should fix the features, can you rebase once it goes in?

**Comment by garious** [comment]
@sakridge, rebased. It's not part of my development process to go debugging CUDA builds. It's great that CI will block this PR, but can you use that failure as a catalyst to add the missing tests to the Rust-only build?

**Comment by garious** [comment]
@sakridge, I'm anxious to get this merged. This change isn't what I wanted to work on; it's merely in the way of me making more important changes to Transaction. Any help getting this one over the line would be much appreciated.

**Comment by garious** [comment]
@sakridge, I think I see the bug. I'm indexing the packet by current_offset.

**Pull request merged**